identity_document/src/document/core_document.rs

```diff
@@ -148,29 +144,106 @@ impl<D: DID + KeyComparable, T, U, V> CoreDocumentData<D, T, U, V> {
 
     Ok(())
   }
+
+  // Apply the provided fallible functions to the DID components of `id`, `controller`, methods and services
+  // respectively.
+  fn try_update_identifiers<F, G, H, L, E>(
```

Comment:
```suggestion
  fn try_map<F, G, H, L, E>(
```

- Can you elaborate why `controller_map`, `method_map` and `services_map` are needed and cannot be the same? Is applying different mappings to methods and services a common enough use case to warrant this method allowing for it or should this method rather do just the simplest case of mapping all contained DIDs in the same way?
- Just to make sure: Can this produce invalid results like the former `try_map` could, e.g. produce methods referencing embedded method identifiers, or services with identifiers matching method identifiers? Not really, right? Because we're mapping only the DID not the DIDUrl?

Reply:
> Can you elaborate why controller_map, method_map and services_map are needed and cannot be the same? 

In `identity_iota_core::StateMetadataDocument::into_iota_document` this method is called with different maps for `controller_map` (because it does extra checks) than `method_map`. In all of our use-cases `method_map` is always the same as `services_map` however, but there might be methods where this is not the case.  We cannot accommodate for every imaginable use case with a method like this, but I don't think it hurts to include this level of flexibility. 

> Just to make sure: Can this produce invalid results like the former try_map could,

Yes it can (although that would be very much an edge case): If you have a document with a service identified by `did:foo:123#example` and a method identified by `did:bar:123#example` and then you map `did:foo -> did:baz` and also `did:bar -> did:baz`.

Reply:
Fair enough, I'm fine with leaving it as-is (except the name, if you agree :smile:).